src/anonymeter/evaluators/singling_out_evaluator.py

```diff
@@ -2,129 +2,168 @@
 
-    return " & ".join(query)
+    expr = reduce(operator.and_, expr_components)
+    return expr
 
 
-def _random_operator(data_type: str) -> str:
-    if data_type == "categorical":
-        ops = ["==", "!="]
-    elif data_type == "boolean":
-        ops = ["", "not "]
+def _random_operator(
+    data_type: str, rng: np.random.Generator
+) -> Callable[[Any, Any], bool]:
+    if data_type in ["categorical", "boolean"]:
+        ops: Sequence[Callable[[Any, Any], bool]] = [operator.eq, operator.ne]
     elif data_type == "numerical":
-        ops = ["==", "!=", ">", "<", ">=", "<="]
+        ops = [
+            operator.eq,
+            operator.ne,
+            operator.gt,
+            operator.lt,
+            operator.ge,
+            operator.le,
+        ]
     else:
         raise ValueError(f"Unknown `data_type`: {data_type}")
 
-    return rng.choice(ops)
-
+    return rng.choice(ops) #type: ignore[arg-type] # signature of "choice" does not accept a list of callables but works fine in practice
 
-def _random_query(unique_values: Dict[str, List[Any]], cols: List[str]):
-    """Generate a random query using given columns."""
-    query = []
 
+def _random_query(
+    unique_values: Dict[str, List[Any]],
+    cols: List[str],
+    column_types: Dict[str, str],
+    rng: Optional[np.random.Generator] = None,
+) -> pl.Expr:
+    rng_to_use = rng_default if rng is None else rng
+    exprs = []
     for col in sorted(cols):
         values = unique_values[col]
-        val = rng.choice(values)
-
-        if pd.isna(val):
-            expression = f"{_random_operator('boolean')}{col}.isna()"
-        elif is_bool_dtype(values):
-            expression = f"{_random_operator('boolean')}{col}"
-        elif isinstance(values, pd.CategoricalDtype):
-            expression = f"{col} {_random_operator('categorical')} {val}"
-        elif is_numeric_dtype(values):
-            expression = f"{col} {_random_operator('numerical')} {val}"
-        elif isinstance(val, str):
-            expression = f"{col} {_random_operator('categorical')} '{_escape_quotes(val)}'"
+        val = rng_default.choice(values)
+
+        data_type = column_types[col]
+        op = _random_operator(data_type, rng_to_use)
+
+        if val is None:
+            # Null checks
+            if op == operator.eq:
+                e = pl.col(col).is_null()
+            else:
+                e = ~pl.col(col).is_null()
+        elif data_type == "boolean":
+            if op == operator.eq:
+                e = pl.col(col)
+            else:
+                e = ~pl.col(col)
         else:
-            expression = f"{col} {_random_operator('categorical')} '{val}'"
+            e = op(pl.col(col), val) #type: ignore[assignment] # mypy does not understand that applying "op" to a polars expression returns a polars expression
 
-        query.append(expression)
+        exprs.append(e)
 
-    return " & ".join(query)
+    return reduce(operator.and_, exprs)
```

Comment:
```suggestion
   # use bitwise and
    return reduce(operator.and_, exprs)
```

if I understand correctly.